spirv_reflect.cpp

```diff
@@ -0,0 +1,514 @@
+            CHECK_MODE(OutputPoints);
+            CHECK_MODE(OutputLineStrip);
+            CHECK_MODE(OutputTriangleStrip);
+            CHECK_MODE(VecTypeHint);
+            CHECK_MODE(ContractionOff);
+
+        default:
+            break;
+        }
+    });
+#endif
+}
+
+void CompilerReflection::emit_resources()
+{
+	auto res = get_shader_resources();
+	emit_resources("subpass_inputs", res.subpass_inputs);
+	emit_resources("inputs", res.stage_inputs);
+	emit_resources("outputs", res.stage_outputs);
+	emit_resources("textures", res.sampled_images);
+	emit_resources("separate_images", res.separate_images);
+	emit_resources("separate_samplers", res.separate_samplers);
+	emit_resources("images", res.storage_images);
+	emit_resources("ssbos", res.storage_buffers);
+	emit_resources("ubos", res.uniform_buffers);
+	emit_resources("push", res.push_constant_buffers);
+	emit_resources("counters", res.atomic_counters);
+}
+
+void CompilerReflection::emit_resources(const char *tag, const vector<Resource> &resources)
+{
+	if (resources.empty())
+	{
+		return;
+	}
+
+	bool print_ssbo = !strcmp(tag, "ssbos");
+
+	emit_json_key_array(tag);
+	for (auto &res : resources)
+	{
+		auto qualifiers = to_qualifiers_glsl(res.id);
+		auto &type = get_type(res.type_id);
+		auto typeflags = meta[type.self].decoration.decoration_flags;
+		auto &dec = meta[type.self].decoration;
+		auto &mask = get_decoration_bitset(res.id);
+		if (print_ssbo && buffer_is_hlsl_counter_buffer(res.id))
+			continue;
+
+		// If we don't have a name, use the fallback for the type instead of the variable
+		// for SSBOs and UBOs since those are the only meaningful names to use externally.
+		// Push constant blocks are still accessed by name and not block name, even though they are technically Blocks.
+		bool is_push_constant = get_storage_class(res.id) == StorageClassPushConstant;
+		bool is_block = get_decoration_bitset(type.self).get(DecorationBlock) ||
+		                get_decoration_bitset(type.self).get(DecorationBufferBlock);
+		uint32_t fallback_id = !is_push_constant && is_block ? res.base_type_id : res.id;
+		begin_json_object();
+		emit_json_key_value("type", type_to_glsl(type));
+		emit_json_key_value("name", !res.name.empty() ? res.name : get_fallback_name(fallback_id));
+		{
+			bool push_constant_block = type.storage == StorageClassPushConstant;
+			bool ssbo_block = type.storage == StorageClassStorageBuffer ||
+			                  (type.storage == StorageClassUniform && typeflags.get(DecorationBufferBlock));
+			if (type.storage == StorageClassUniform && typeflags.get(DecorationBlock))
+			{
+				if (buffer_is_packing_standard(type, BufferPackingStd140))
+					emit_json_key_value("std140", true);
+				else if (buffer_is_packing_standard(type, BufferPackingStd140EnhancedLayout))
+					emit_json_key_value("std140", true);
+			}
+			else if (push_constant_block || ssbo_block)
+			{
+				if (buffer_is_packing_standard(type, BufferPackingStd430))
```

Comment:
I don't think this is relevant for SPIR-V reflection.
std430/std140 only make sense for GLSL output/input.

Reply:
If someone intends to use the reflection in combination with GLSL output from SPIRV-Cross, then it's critical, is it not?  That's actually my use case.  This is part of a first step towards a full Vulkan backend, by making all the shaders SPIRV-Compatibile and using SPIRV as our transmission format, even though we _might_ have to use GLSL for the shader upload to the GPU on systems that aren't OpenGL 4.6 or higher.

Reply:
Well, the question then is, the API-side of GL should not care about packing used in the shader (unless it's packed/shared layout). What information does your backend use if a buffer is std140 or std430? I don't think that can even be queried. The shader finds whatever layout will conform to the SPIR-V, not the other way around.

Reply:
That makes sense.